dpnp/backend/extensions/indexing/choose.cpp

```diff
@@ -0,0 +1,453 @@
+        throw std::runtime_error("Indices must be integer type, got " +
+                                 std::to_string(src_type_id));
+    }
+
+    auto packed_chc_ptrs = usm_unique_ptr<char *>(n_chcs, exec_q);
+    if (packed_chc_ptrs.get() == nullptr) {
+        throw std::runtime_error(
+            "Unable to allocate packed_chc_ptrs device memory");
+    }
+
+    // packed_shapes_strides = [common shape,
+    //                          src.strides,
+    //                          dst.strides,
+    //                          chcs[0].strides,
+    //                          ...,
+    //                          chcs[n_chcs].strides]
+    auto packed_shapes_strides =
+        usm_unique_ptr<py::ssize_t>((3 + n_chcs) * sh_nelems, exec_q);
+    if (packed_shapes_strides.get() == nullptr) {
+        throw std::runtime_error(
+            "Unable to allocate packed_shapes_strides device memory");
+    }
+
+    auto packed_chc_offsets = usm_unique_ptr<py::ssize_t>(n_chcs, exec_q);
+    if (packed_chc_offsets.get() == nullptr) {
+        throw std::runtime_error(
+            "Unable to allocate packed_chc_offsets device memory");
+    }
+
+    std::vector<sycl::event> host_task_events;
+    host_task_events.reserve(2);
+
+    std::vector<sycl::event> pack_deps;
+    if (nd == 0) {
+        // special case where all inputs are scalars
+        // need to pass src, dst shape=1 and strides=0
+        // chc_strides already initialized to 0 so ignore
+        std::array<py::ssize_t, 1> scalar_sh{1};
+        std::vector<py::ssize_t> src_strides{0};
+        std::vector<py::ssize_t> dst_strides{0};
+
+        pack_deps = _populate_choose_kernel_params(
+            exec_q, host_task_events, packed_chc_ptrs.get(),
+            packed_shapes_strides.get(), packed_chc_offsets.get(),
+            scalar_sh.data(), sh_nelems, src_strides, dst_strides, chc_strides,
+            chc_ptrs, chc_offsets, n_chcs);
+    }
+    else {
+        auto src_strides = src.get_strides_vector();
+        auto dst_strides = dst.get_strides_vector();
+
+        pack_deps = _populate_choose_kernel_params(
+            exec_q, host_task_events, packed_chc_ptrs.get(),
+            packed_shapes_strides.get(), packed_chc_offsets.get(), src_shape,
+            sh_nelems, src_strides, dst_strides, chc_strides, chc_ptrs,
+            chc_offsets, n_chcs);
+    }
+
+    std::vector<sycl::event> all_deps;
+    all_deps.reserve(depends.size() + pack_deps.size());
+    all_deps.insert(std::end(all_deps), std::begin(pack_deps),
+                    std::end(pack_deps));
+    all_deps.insert(std::end(all_deps), std::begin(depends), std::end(depends));
+
+    sycl::event choose_generic_ev =
+        fn(exec_q, nelems, n_chcs, sh_nelems, packed_shapes_strides.get(),
+           src_data, dst_data, packed_chc_ptrs.get(), src_offset, dst_offset,
+           packed_chc_offsets.get(), all_deps);
+
+    // release usm_unique_ptrs
+    auto chc_ptrs_ = packed_chc_ptrs.release();
```

Comment:
After some thoughts I think it would be easier just to use `shared_ptr`. 
I don't like `shared_ptr` due its overhead, but it is not critical.

Alternatively we can introduce some utilities for `unique_ptr`.
To do it in the following way:
```c++
auto temporaries_cleanup_ev = async_free(
    exec_q, 
    choose_generic_ev, 
    std::move(packed_chc_ptrs), 
    std::move(...),
    ...);
```

I have an example implementation of `async_free` if you wish to go this way.

Anyway it is better to stick one common way - either `unique_ptr` or `shared_ptr` both here and in `_populate_choose_kernel_params` function.